### PR TITLE
fill-pattern with pixelRatio > 1 is now switched correctly at runtime

### PR DESCRIPTION
**Pull request by zhangyiatmicrosoft**
## Launch Checklist

Fix issue https://github.com/maplibre/maplibre-gl-js/issues/860 : if image is created with pixelRatio > 1, use it at run time as "fill-pattern" will cause flash.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [x] Post benchmark scores.
 - [x] Manually test the debug page.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by zhangyiatmicrosoft** [comment]
Before 
https://1drv.ms/u/s!As5idA_oj10Xga0ASZWvotK2fRa8CA?e=9Fm4Tu
After:
https://1drv.ms/u/s!As5idA_oj10Xgax_82dHvKm51A4-SA?e=wHLUrp

**Comment by zhangyiatmicrosoft** [comment]
This is affecting transitioning frames only and I don't see any way to capture them and test automatically.
The best I can do is to add a debug page for manual testing. Please let me know if more can be done

**Comment by zhangyiatmicrosoft** [comment]
Before: https://1drv.ms/u/s!As5idA_oj10Xga0DADHBUzwwezvM9w?e=IfDCdx
After: https://1drv.ms/v/s!As5idA_oj10Xga0CbhL1706tjWG6oQ?e=MVwB94

**Comment by HarelM** [comment]
I don't have a good idea how to test this in terms of how it would look, having said that I think that if you create a new method it will be best to move it to a new file instead of exporting two method in the same file.
This new file should be unit tested to cover the case here and to prevent this from breaking in the future.
Also list is failing :-)
Thanks for taking the time to solve this!

**Comment by zhangyiatmicrosoft** [comment]
Benchmark results.

image

**Comment by zhangyiatmicrosoft** [comment]
> I don't have a good idea how to test this in terms of how it would look, having said that I think that if you create a new method it will be best to move it to a new file instead of exporting two method in the same file. This new file should be unit tested to cover the case here and to prevent this from breaking in the future. Also list is failing :-) Thanks for taking the time to solve this!

Thanks. Added draw_fill_test and factored the new function out to its own file

File: src/render/draw_fill_helper.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import type {CrossFaded} from '../style/properties';
```
Comment (HarelM) [inline src/render/draw_fill_helper.ts:1]: I would keep the file name and the exported method the same - `find_pattern_positions.ts`

File: src/render/draw_fill_helper.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import type ResolvedImage from '../style-spec/expression/types/resolved_image';
+import type Tile from '../source/tile';
+import type ProgramConfiguration from '../data/program_configuration';
+import type FillExtrusionStyleLayer from '../style/style_layer/fill_extrusion_style_layer';
+import type FillStyleLayer from '../style/style_layer/fill_style_layer';
+
+export default findPatternPositions;
```
Comment (HarelM) [inline src/render/draw_fill_helper.ts:8]: You can use `export default function findPatternPositions(...) { ... }` and reduce this extra line.

File: src/render/draw_fill_helper.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * @param programConfiguration - to be used to set patttern poistion and device pixel ratio.
+ */
+function findPatternPositions(
+    propertyName: 'fill-pattern' | 'fill-extrusion-pattern',
+    constantPattern: CrossFaded<ResolvedImage>,
+    tile: Tile,
+    layer: FillStyleLayer | FillExtrusionStyleLayer,
+    programConfiguration: ProgramConfiguration): void {
+
+    if (constantPattern && tile.imageAtlas) {
```
Comment (HarelM) [inline src/render/draw_fill_helper.ts:30]: can we invert this if to reduce indentation?

File: src/render/draw_fill_helper.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * For transtionable properties, especially 'fill-pattern' and 'fill-extrusion-pattern', at certain frames
+ * tile.imageAtlas has been updated by worker to holds the new pattern only, but code here is still looking for previous images,
+ * causing a few corrupted frames in which setConstantPatternPositions method is not called and pixelRatio is always the
+ * default of 1, instead of actual values set by original map.addImage
+ *
+ * @param propertyName - 'fill-pattern' or 'fill-extrusion-pattern' property key
+ * @param constantPattern - either 'fill-pattern' or 'fill-extrusion-pattern' property value
+ * @param tile - current tile being drawn
+ * @param layer - current layer being rendered
+ * @param programConfiguration - to be used to set patttern poistion and device pixel ratio.
```
Comment (HarelM) [inline src/render/draw_fill_helper.ts:21]: The documentation here is generally great, but it's not clear that the method is updating something.
The method name is `findPatternPositions` I would expect this function to return the position or something similar, but instead it updates the program. I would consider changing the name to something like `updatePatternPositionsInProgram` and have the program configuration as the first parameter.

File: src/render/draw_fill.test.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+jest.mock('./painter');
+jest.mock('./program');
+jest.mock('../source/source_cache');
+jest.mock('../source/tile');
+jest.mock('../data/bucket/symbol_bucket');
+jest.mock('../symbol/projection');
+
+describe('drawSymbol', () => {
+    test('should call programConfiguration.setConstantPatternPositions for transitioning fill-pattern', () => {
```
Comment (HarelM) [inline src/render/draw_fill.test.ts:26]: This test is very good, I would consider adding a test file to the added file to reduce the need for a lot of mock to test the code you added.
Having said that, this test is very good too. :-)

Comment (zhangyiatmicrosoft) [reply src/render/draw_fill.test.ts:26]: That was my initial thought too but then I realized to test the new file I still need 4 mocks.
With 6 mocks, the current one follows the same idea as draw_symbol.test, adds coverage to draw_fill, and tests the new file at same time, so it is much bigger return with slightly increased investment. :)
I feel probably don't need another test to repeat things already covered? :)

BTW, just refactored the code a little to make it easier to read and add more test cases when needed. Thanks

**Review by HarelM — APPROVED** [review]
Thanks for taking the time to dig into this!

**Pull request merged**